cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataKey.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.data2.metadata.dataset;
+
+import co.cask.cdap.api.metadata.MetadataEntity;
+import co.cask.cdap.data2.dataset2.lib.table.EntityIdKeyHelper;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.id.NamespacedEntityId;
+
+import java.nio.BufferUnderflowException;
+import javax.annotation.Nullable;
+
+/**
+ * Key used to store Metadata values and indexes
+ */
+class MetadataKey {
+  private static final byte[] VALUE_ROW_PREFIX = {'v'}; // value row prefix to store metadata value
+  private static final byte[] INDEX_ROW_PREFIX = {'i'}; // index row prefix used for metadata search
+  // if a target type cannot be determined because the MetadataEntity is an unknown CDAP entity then we will store it
+  // as CUSTOM_TYPE.
+  private static final String CUSTOM_TARGET_TYPE = "CUSTOM_TYPE";
+
+  static String getMetadataKey(byte[] rowKey) {
+    MDSKey.Splitter keySplitter = new MDSKey(rowKey).split();
+    // The rowkey is
+    // [rowPrefix][targetType][targetId][key] for value rows and
+    // [rowPrefix][targetType][targetId][key][index] for value index rows
+
+    // Skip rowPrefix
+    keySplitter.skipBytes();
+    // Skip targetType
+    keySplitter.skipString();
+
+    // targetId are key-value par so always in set of two. For value row we will end up with only string in end ([key])
+    // and for index row we will have two strings in end ([key][index]). In the first case trying to read second
+    // string to lead BufferUnderFlowException which we catch and ignore returning the key.
+    String key = null;
+    while (keySplitter.hasRemaining()) {
+      key = keySplitter.getString();
```

Comment:
You should still rename KeySplitter.getString() to readString(), because it is not a getter, it actually mutates the state of the splitter.  Any method called get...() should be idempotent.